tests.py

```diff
@@ -1,107 +1,192 @@
+from segmentation import MatrixVariateNormalInvGamma
 
+
+def random_segments(m, n, k, l, mu=None, omega=None, sigma=None, eta=None,
+                    featfun=None):
+    """Generate random segmented multi-variate linear model data."""
+
+    # Set uninformative prior for the location parameter.
+    if mu is None:
+        mu = np.zeros([m, n])
+
+    # Set uninformative prior for the scale parameter.
+    if omega is None:
+        omega = np.eye(m)
+
+    # Set uninformative prior for the dispersion/noise parameter.
+    if sigma is None:
+        sigma = np.eye(n)
+
+    # Set uninformative prior for the shape parameter.
+    if eta is None:
+        eta = n
+
+    featfun = featfun if callable(featfun) else lambda x: x
+
+    # Generate the segment boundaries.
+    bound = random.permutation(np.arange(k - 1) + 1)
+    bound = np.concatenate([np.array([0]),
+                            np.sort(bound[:l-1]),
+                            np.array([k])])
+
+    # For each segment generate a set of predictor-response data.
+    X, Y = list(), list()
+    for i in range(len(bound) - 1):
+
+        # Generate random predictor (input) data and pre-allocate memory for
+        # response (output) data.
+        k = bound[i + 1] - bound[i]
+        x = random.rand(k, m)
+        y = np.zeros((k, n))
+
+        # Generate the gain and noise parameters.
+        gain, noise = MatrixVariateNormalInvGamma(mu, omega, sigma, eta).rand()
+        fact = linalg.cholesky(noise).transpose()
+
+        # Given a set of predictor data, generate a corresponding set of
+        # response data.
+        for j in range(k):
+            y[j, :] = featfun(np.dot(x[j, :], gain))
+            y[j, :] += np.dot(random.randn(n), fact)
+
+        X.append(x)
+        Y.append(y)
+
+    # Adjust the last element of the true boundaries for python's zero
+    # indexing.
+    bound[-1] -= 1
+
+    return bound, np.concatenate(X, axis=0), np.concatenate(Y, axis=0)
+
+
+def plot_probability(axes, prob, scale=None, **arg):
+
+    """Plot hypotheses probability as a raster."""
+
+    if scale is None:
+        scale = lambda x: x
+
+    k = max(np.shape(prob)) - 1
+    ind, = np.nonzero(prob.max(axis=1) > 0)
+    j = ind.max()
+
+    # Plot the posterior probabilities of the segmentation hypotheses.
+    axes.imshow(1.0 - prob[:j+1],
+                origin='lower',
+                aspect='auto',
+                extent=[scale(-0.5), scale(k + 0.5), -0.5, j + 0.5],
+                interpolation='none',
+                **arg)
+
+
+def plot_segment_span(x, segments, **arg):
+    """Plot segments as alternating vertical spans (rectangles)."""
+
+    for i in range(len(segments) - 1):
+        if i % 2 == 0:
+            plt.axvspan(x[segments[i]], x[segments[i + 1]], **arg)
+
+
+def plot_segment_boundaries(x, segments, **args):
+    """Plot segment boundaries as vertical lines."""
+
+    for i in range(len(segments)):
+        plt.axvline(x[segments[i]], **args)
```

Comment:
The matplotlib functions also plot to infinity which makes panning around the figure nicer.